outwatch/src/main/scala/outwatch/util/Store.scala

```diff
@@ -14,58 +14,63 @@ import scala.util.control.NonFatal
 
 
 object Store {
+  sealed trait BaseAction
+  private final case object Noop extends BaseAction
+  private final case class Error(context: Action, error: Throwable) extends BaseAction
+  trait Action extends BaseAction
 
-  case class Reducer[S, A](reducer: (S, A) => (S, Observable[A]))
+  case class Reducer[A, M](reducer: (M, A) => (M, Observable[A]))
 
   object Reducer {
+    implicit def stateAndEffects[A, M](f: (M, A) => (M, Observable[A])): Reducer[A, M] = Reducer(f)
 
-    implicit def stateAndEffects[S, A](f: (S, A) => (S, Observable[A])): Reducer[S, A] = Reducer(f)
+    implicit def justState[A, M](f: (M, A) => M): Reducer[A, M] = Reducer { (s: M, a: A) => (f(s, a), Observable.empty) }
 
-    implicit def justState[S, A](f: (S, A) => S): Reducer[S, A] = Reducer { (s: S, a: A) => (f(s, a), Observable.empty) }
-
-    implicit def stateAndOptionIO[S, A](f: (S, A) => (S, Option[IO[A]])): Reducer[S, A] =  Reducer { (s: S, a: A) =>
-      val (mewState, effect) = f(s, a)
-      (mewState, effect.fold[Observable[A]](Observable.empty)(Observable.fromIO))
+    implicit def stateAndOptionIO[A, M](f: (M, A) => (M, Option[IO[A]])): Reducer[A, M] =  Reducer { (s: M, a: A) =>
+      val (newState, effect) = f(s, a)
+      (newState, effect.fold[Observable[A]](Observable.empty)(Observable.fromIO))
     }
   }
 
-  private val storeRef = STRef.empty
-
-  def create[State, Action](
-    initialState: State,
-    reducer: Reducer[State, Action]
-  )(implicit s: Scheduler): IO[ProHandler[Action, State]] = IO {
-
-      val subject = PublishSubject[Action]
-
-      val fold: (State, Action) => State = (state, action) => Try { // guard against reducer throwing an exception
-        val (newState, effects) = reducer.reducer(state, action)
-
-        effects.subscribe(
-          e => subject.feed(e :: Nil),
-          e => dom.console.error(e.getMessage) // just log the error, don't push it into the subject's observable, because it would stop the scan "loop"
-        )
-        newState
-      }.recover { case NonFatal(e) =>
-        dom.console.error(e.getMessage)
-        state
-      }.get
+  def create[A <: Action, M](
+    initialState: M,
+    reducer: Reducer[A, M]
+  )(implicit s: Scheduler): IO[ProHandler[A, (BaseAction, M)]] = IO {
+      val subject = PublishSubject[A]
+
+      val fold: ((BaseAction, M), A) => (BaseAction, M) = {
+        case ((_, state), action) => Try { // guard against reducer throwing an exception
+          val (newState, effects) = reducer.reducer(state, action)
+
+          effects.subscribe(
+            next => subject.feed(next :: Nil),
+            // just log the error, don't push it into the subject's observable, because it would stop the scan "loop"
+            error => dom.console.error(error.getMessage)
```

Comment:
Perhaps more flexible to use an `ErrorSink` instead `dom.console.error`, or at least have a way to turn logging off.

Reply:
Good idea! But it does not necessarily need to be fixed in this PR.

Reply:
Just a thought :)